src/threading/MsgThread.cc

```diff
@@ -165,6 +165,72 @@ bool ReporterMessage::Process() {
     return true;
 }
 
+
+// This is the IO source used by MsgThread.
+//
+// The lifetime of the IO source is decoupled from
+// the thread. The thread may be terminated prior
+// to the IO source being properly unregistered and
+// forgotten by the IO manager. Specifically the
+// threading manager would delete an IO source which
+// the IO manager still believed to be ready.
+//
+// See issue #3682 for more details.
+class MsgThread_IOSource : public iosource::IOSource {
+public:
+    explicit MsgThread_IOSource(MsgThread* thread) : thread(thread) {
+        if ( ! iosource_mgr->RegisterFd(flare.FD(), this) )
+            reporter->FatalError("Failed to register MsgThread fd with iosource_mgr");
```

Comment:
`FatalError` a user-level error message. I'd either rephrase it to make it at least somewhat better understandable by users, or, if it's something that really shouldn't happen, make it an `InternalError` instead. More of these elsewhere.